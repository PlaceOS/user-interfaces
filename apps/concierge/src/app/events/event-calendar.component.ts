import { Component, Input } from '@angular/core';
import { SettingsService } from '@placeos/common';
import { EventStateService } from './event-state.service';
import { map, shareReplay, take } from 'rxjs/operators';
import { addMonths, addWeeks, format } from 'date-fns';
import { ActivatedRoute, Router } from '@angular/router';

@Component({
    selector: 'event-calendar',
    template: `
        <div class="flex items-center border-y border-base-200 w-full">
            <div class="flex-1 px-2 py-4">
                <span *ngIf="is_today | async" class="text-info text-xs">{{
                    (period | async) === 'week' ? 'This Week' : 'This Month'
                }}</span>
            </div>
            <div class="flex-2 flex items-center justify-center space-x-2">
                <div class="font-medium pl-4">
                    {{ (options | async)?.date | date: 'MMM yyyy' }}
                </div>
                <button icon matRipple (click)="previousPeriod()">
                    <app-icon>chevron_left</app-icon>
                </button>
                <button icon matRipple (click)="nextPeriod()">
                    <app-icon>chevron_right</app-icon>
                </button>
            </div>
            <div class="flex-1 flex items-center justify-end space-x-2 py-2">
                <button
                    btn
                    matRipple
                    class="rounded-3xl"
                    [class.inverse]="(period | async) !== 'week'"
                    (click)="setPeriod('week')"
                >
                    Week
                </button>
                <button
                    btn
                    matRipple
                    class="rounded-3xl"
                    [class.inverse]="(period | async) !== 'month'"
                    (click)="setPeriod('month')"
                >
                    Month
                </button>
            </div>
        </div>
        <div class="w-full h-4"></div>
        <div class="relative flex-1 h-1/2 w-full overflow-auto">
            <event-week-view
                *ngIf="(period | async) !== 'month'; else month_calendar"
            ></event-week-view>
            <ng-template #month_calendar>
                <event-month-view></event-month-view>
            </ng-template>
        </div>
    `,
    styles: [
        `
            :host {
                display: flex;
                flex-direction: column;
                height: 100%;
                width: 100%;
            }
        `,
    ],
})
export class EventCalendarComponent {
    public readonly period = this._state.options.pipe(map((_) => _.period));

    public readonly options = this._state.options;
    public readonly is_today = this.options.pipe(
        map((_) => _.date <= Date.now() && _.end > Date.now())
    );
    public readonly event_list = this._state.event_list;
    public readonly event_day_map = this.event_list.pipe(
        map((list) => {
            const map = {};
            for (const event of list) {
                const date = format(event.date, 'yyyy-MM-dd');
                if (!map[date]) map[date] = [];
                map[date].push(event);
            }
            return map;
        }),
        shareReplay(1)
    );

    public get time_format() {
        return this._settings.time_format;
    }

    constructor(
        private _settings: SettingsService,
        private _state: EventStateService,
        private _router: Router,
        private _route: ActivatedRoute
    ) {}

    public setPeriod(period: 'week' | 'month') {
        this._state.setOptions({ period });
        this._router.navigate([], {
            relativeTo: this._route,
            queryParams: { period },
            queryParamsHandling: 'merge',
        });
    }

    public async nextPeriod() {
        const { date, period } = await this.options.pipe(take(1)).toPromise();
        this._router.navigate([], {
            relativeTo: this._route,
            queryParams: {
                range:
                    period === 'week'
                        ? addWeeks(date, 1).valueOf()
                        : addMonths(date, 1).valueOf(),
            },
            queryParamsHandling: 'merge',
        });
    }

    public async previousPeriod() {
        const { date, period } = await this.options.pipe(take(1)).toPromise();
        this._router.navigate([], {
            relativeTo: this._route,
            queryParams: {
                range:
                    period === 'week'
                        ? addWeeks(date, -1).valueOf()
                        : addMonths(date, -1).valueOf(),
            },
            queryParamsHandling: 'merge',
        });
    }
}
