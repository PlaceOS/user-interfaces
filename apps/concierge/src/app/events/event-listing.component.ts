import { Component } from '@angular/core';
import { SettingsService } from '@placeos/common';
import { EventStateService } from './event-state.service';
import { User } from '@placeos/users';
import { tap } from 'rxjs/operators';

@Component({
    selector: 'event-listing',
    template: `
        <simple-table
            class="min-w-[72rem] w-full block text-sm"
            [data]="event_list"
            empty_message="No Regions"
            [columns]="[
                { key: 'date', name: 'Event', content: event_template },
                {
                    key: 'level',
                    name: 'Level',
                    content: level_template,
                    size: '8rem',
                    sortable: false
                },
                {
                    key: 'room',
                    name: 'Room',
                    content: room_template,
                    size: '12rem',
                    sortable: false
                },
                {
                    key: 'interested',
                    name: 'Interested',
                    content: interested_template,
                    size: '6rem',
                    sortable: false
                },
                {
                    key: 'attending',
                    name: 'Attending',
                    content: attending_template,
                    size: '6rem',
                    sortable: false
                },
                {
                    key: 'state',
                    name: 'Status',
                    content: status_template,
                    size: '8.5rem'
                },
                {
                    key: 'permissions',
                    name: 'Published',
                    content: published_template,
                    size: '6rem',
                    sortable: false
                },
                {
                    key: 'actions',
                    name: ' ',
                    content: actions_template,
                    size: '4rem',
                    sortable: false
                }
            ]"
            [sortable]="true"
        ></simple-table>
        <div class="w-full h-20"></div>
        <ng-template #event_template let-item="row">
            <div class="flex items-center space-x-2 px-3 py-2">
                <div date class="flex flex-col items-center leading-tight w-8">
                    <div
                        month
                        class="text-sm font-medium relative top-0.5 opacity-60"
                    >
                        {{ item.date | date: 'MMM' }}
                    </div>
                    <div day class="text-2xl font-light relative -top-0.5">
                        {{ item.date | date: 'd' }}
                    </div>
                </div>
                <div
                    class="flex items-center justify-center h-12 w-12 rounded overflow-hidden bg-base-200 border border-base-200"
                >
                    <img
                        *ngIf="item.images?.length"
                        auth
                        [source]="item.images[0]"
                        class="min-h-full min-w-full object-cover"
                    />
                </div>
                <div details class="flex flex-col">
                    <div class="text-sm">{{ item.title }}</div>
                    <div class="text-xs opacity-40">
                        {{ item.date | date: 'EEEE, ' + time_format }} &ndash;
                        {{
                            item.date + item.duration * 60 * 1000
                                | date: time_format + ' (z)'
                        }}
                    </div>
                    <div class="text-xs">
                        {{
                            (
                                (item.linked_event?.system_id | space | async)
                                    ?.zones | building
                            )?.address
                        }}
                    </div>
                </div>
            </div>
        </ng-template>
        <ng-template #level_template let-item="row">
            <div class="p-4">
                {{
                    (
                        (item.linked_event?.system_id | space | async)?.zones
                        | level
                    )?.display_name
                }}
                <span *ngIf="!item.linked_event?.system_id" class="opacity-30">
                    No Level
                </span>
            </div>
        </ng-template>
        <ng-template #room_template let-item="row">
            <div class="p-4">
                {{
                    (item.linked_event?.system_id | space | async)?.display_name
                }}
                <span *ngIf="!item.linked_event?.system_id" class="opacity-30">
                    No Room
                </span>
            </div>
        </ng-template>
        <ng-template #interested_template let-item="row">
            <button
                icon
                matRipple
                customTooltip
                [content]="view_attendees"
                class="rounded h-12 w-12 mx-auto"
                [disabled]="!item.attendees?.length"
            >
                {{ item.attendees?.length || 0 }}
            </button>
            <ng-template #view_attendees>
                <div
                    class="relative w-[20rem] h-[28rem] overflow-auto bg-white rounded shadow"
                >
                    <attendee-list
                        [list]="item.attendees"
                        [host]="item.user_email"
                        [hide_close]="true"
                        [custom_title]="'Interested'"
                    ></attendee-list>
                </div>
            </ng-template>
        </ng-template>
        <ng-template #attending_template let-item="row">
            <button
                icon
                matRipple
                customTooltip
                [content]="view_attendees"
                class="rounded h-12 w-12 mx-auto"
                [disabled]="!checkedInCount(item.attendees)"
            >
                {{ checkedInCount(item.attendees) }}
            </button>
            <ng-template #view_attendees>
                <div
                    class="relative w-[20rem] h-[28rem] overflow-auto bg-white rounded shadow"
                >
                    <attendee-list
                        [list]="checkedInList(item.attendees)"
                        [host]="item.user_email"
                        [hide_close]="true"
                    ></attendee-list>
                </div>
            </ng-template>
        </ng-template>
        <ng-template #published_template let-row="row">
            <div
                *ngIf="row.permission === 'OPEN' || row.permission === 'open'"
                class="rounded h-8 w-8 flex items-center justify-center text-2xl bg-success text-success-content mx-auto"
            >
                <app-icon>done</app-icon>
            </div>
        </ng-template>
        <ng-template #status_template let-item="row">
            <div class="p-4">
                <div
                    class="px-4 py-1 rounded-full"
                    [class.bg-success]="
                        item.state !== 'done' &&
                        item.state !== 'in_progress' &&
                        item.state !== 'started'
                    "
                    [class.text-success-content]="
                        item.state !== 'done' &&
                        item.state !== 'in_progress' &&
                        item.state !== 'started'
                    "
                    [class.bg-warning]="
                        item.state === 'in_progress' || item.state === 'started'
                    "
                    [class.text-warning-content]="
                        item.state === 'in_progress' || item.state === 'started'
                    "
                    [class.bg-base-200]="item.state === 'done'"
                    [class.text-base-content]="item.state === 'done'"
                >
                    {{
                        item.state === 'done'
                            ? 'Done'
                            : item.state === 'in_progress' ||
                              item.state === 'started'
                            ? 'In Progress'
                            : 'Active'
                    }}
                </div>
            </div>
        </ng-template>
        <ng-template #actions_template let-row="row">
            <button
                icon
                matRipple
                class="h-12 w-12 rounded mx-2"
                [matMenuTriggerFor]="menu"
                [disabled]="row.state === 'done'"
            >
                <app-icon>more_vert</app-icon>
            </button>
            <mat-menu #menu="matMenu">
                <button mat-menu-item [disabled]="true">
                    <div class="flex items-center space-x-2">
                        <app-icon class="text-2xl">
                            confirmation_number
                        </app-icon>
                        <div class="mr-2">Promote Event</div>
                    </div>
                </button>
                <button mat-menu-item (click)="viewEvent(row)">
                    <div class="flex items-center space-x-2">
                        <app-icon class="text-2xl">visibility</app-icon>
                        <div class="mr-2">View Event</div>
                    </div>
                </button>
                <a
                    mat-menu-item
                    [routerLink]="[
                        '/entertainment',
                        'events',
                        'manage',
                        row?.id
                    ]"
                >
                    <div class="flex items-center space-x-2">
                        <app-icon class="text-2xl">edit</app-icon>
                        <div class="mr-2">Edit Event</div>
                    </div>
                </a>
                <button mat-menu-item [disabled]="true">
                    <div class="flex items-center space-x-2">
                        <app-icon class="text-2xl">content_copy</app-icon>
                        <div class="mr-2">Copy URL</div>
                    </div>
                </button>
                <button mat-menu-item (click)="removeEvent(row)">
                    <div class="flex items-center space-x-2">
                        <app-icon class="text-2xl text-error">delete</app-icon>
                        <div class="mr-2">Delete Event</div>
                    </div>
                </button>
            </mat-menu>
        </ng-template>
    `,
    styles: [``],
})
export class EventListingComponent {
    public readonly event_list = this._state.event_list.pipe(
        tap((_) => console.log('Event List:', _))
    );

    public readonly viewEvent = (event: any) => this._state.viewEvent(event);
    public readonly removeEvent = (event: any) =>
        this._state.removeEvent(event);

    public get time_format() {
        return this._settings.time_format;
    }

    constructor(
        private _settings: SettingsService,
        private _state: EventStateService
    ) {}

    public checkedInCount(attendees: User[]) {
        if (!attendees?.length) return 0;
        return attendees.filter((user: User) => user.checked_in).length;
    }

    public checkedInList(attendees: User[]) {
        if (!attendees?.length) return [];
        return attendees.filter((user: User) => user.checked_in);
    }
}
