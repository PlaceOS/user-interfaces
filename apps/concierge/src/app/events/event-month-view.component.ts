import { Component } from '@angular/core';
import { EventStateService } from './event-state.service';
import { AsyncHandler, SettingsService } from '@placeos/common';
import {
    format,
    addDays,
    isSameDay,
    isSameMonth,
    startOfDay,
    startOfMonth,
    startOfWeek,
} from 'date-fns';
import { map, shareReplay, startWith } from 'rxjs/operators';
import { MatDialog } from '@angular/material/dialog';
import { Router } from '@angular/router';

@Component({
    selector: 'event-month-view',
    template: `
        <div class="absolute inset-0 overflow-auto">
            <div
                class="grid grid-cols-7 min-w-full w-[80rem] min-h-full h-[56rem] m-2 border-b border-base-200 "
            >
                <div
                    weekday
                    *ngFor="let weekday of weekdays"
                    class="relative flex items-center justify-center text-sm opacity-60 p-2 h-12"
                >
                    {{ weekday | date: 'EEEE' }}
                </div>
                <div
                    monthday
                    *ngFor="let day of month_days"
                    class="relative border border-base-200 flex flex-col space-y-1"
                >
                    <div
                        class="rounded-full w-8 h-8 flex items-center justify-center mt-1 ml-1"
                        [class.opacity-30]="!day.is_month"
                        [class.bg-secondary]="day.is_today"
                        [class.text-secondary-content]="day.is_today"
                    >
                        {{ day.id | date: 'd' }}
                    </div>
                    <button
                        matRipple
                        *ngFor="
                            let event of (event_day_map | async)[
                                dateString(day.id)
                            ] || [] | slice: 0:3
                        "
                        (click)="viewEvent(event)"
                        class="relative w-[calc(100%-0.5rem)] h-7 bg-base-100 rounded border border-base-200 hover:border-info shadow pl-3 pr-2 py-1 overflow-hidden mx-1"
                    >
                        <div
                            class="absolute left-0 inset-y-0 bg-info w-1.5"
                        ></div>
                        <div
                            class="text-sm opacity-60 h-full text-left truncate"
                        >
                            {{ event.date | date: 'shortTime' }} &mdash;
                            {{ event.title }}
                        </div>
                        <div
                            class="absolute inset-0"
                            customTooltip
                            [content]="event_card"
                            xPosition="center"
                            yPosition="bottom"
                            [hover]="true"
                        ></div>
                        <ng-template #event_card>
                            <div class="p-2 pointer-events-none">
                                <group-event-card
                                    [event]="event"
                                ></group-event-card>
                            </div>
                        </ng-template>
                    </button>
                    <button
                        matRipple
                        *ngIf="
                            ((event_day_map | async)[dateString(day.id)] || [])
                                .length > 3
                        "
                        matTooltip="More events"
                        class="relative w-[calc(100%-0.5rem)] h-7 rounded pl-3 pr-2 py-1 overflow-hidden mx-1 underline text-sm"
                        [matMenuTriggerFor]="menu"
                    >
                        {{
                            ((event_day_map | async)[dateString(day.id)] || [])
                                .length - 3
                        }}
                        more event(s)
                    </button>
                    <mat-menu #menu="matMenu">
                        <button
                            mat-menu-item
                            *ngFor="
                                let event of (event_day_map | async)[
                                    dateString(day.id)
                                ] || [] | slice: 3
                            "
                            (click)="viewEvent(event)"
                        >
                            <div class="flex items-center space-x-4">
                                <div class="flex-1">{{ event.title }}</div>
                                <div
                                    class="text-xs opacity-60 py-1 px-2 rounded bg-base-200 text-base-content"
                                >
                                    {{ event.date | date: 'shortTime' }}
                                </div>
                            </div>
                        </button>
                    </mat-menu>
                </div>
            </div>
        </div>
    `,
    styles: [
        `
            .grid {
                grid-template-rows: 3rem repeat(6, minmax(0, 1fr));
            }
        `,
    ],
})
export class EventMonthViewComponent extends AsyncHandler {
    public month = startOfDay(Date.now()).valueOf();
    public weekdays = [];

    public month_days = [];
    public readonly event_list = this._state.event_list;
    public readonly event_day_map = this.event_list.pipe(
        map((list) => {
            const map = {};
            for (const event of list) {
                const date = format(event.date, 'yyyy-MM-dd');
                if (!map[date]) map[date] = [];
                const start = new Date(event.date);
                map[date].push({
                    ...event,
                    offset:
                        (start.getHours() * 60 + start.getMinutes()) /
                        (24 * 60),
                    length: event.duration / (24 * 60),
                });
            }
            return map;
        }),
        startWith({}),
        shareReplay(1)
    );

    public readonly viewEvent = (event: any) => this._state.viewEvent(event);

    public dateString(date: number) {
        if (!date) return '';
        return format(date, 'yyyy-MM-dd');
    }

    public get offset_weekday() {
        return this._settings.get('app.week_start') || 0;
    }

    constructor(
        private _state: EventStateService,
        private _settings: SettingsService,
        private _dialog: MatDialog,
        private _router: Router
    ) {
        super();
    }

    public ngOnInit() {
        this.subscription(
            'date',
            this._state.options.subscribe(({ date }) => {
                this.month = date;
                this._setMonthDays();
                this._setWeekdays();
            })
        );
        this._setMonthDays();
        this._setWeekdays();
    }

    private _setMonthDays() {
        const start = startOfWeek(startOfMonth(this.month), {
            weekStartsOn: this.offset_weekday as any,
        });
        this.month_days = Array.from(Array(7 * 6).keys()).map((i) => {
            const date = addDays(start, i).valueOf();
            return {
                id: date,
                is_today: isSameDay(date, Date.now()),
                is_month: isSameMonth(date, this.month),
            };
        });
    }

    private _setWeekdays() {
        const start = startOfWeek(Date.now(), {
            weekStartsOn: this.offset_weekday as any,
        });
        this.weekdays = Array.from(Array(7).keys()).map((i) =>
            addDays(start, i)
        );
    }
}
