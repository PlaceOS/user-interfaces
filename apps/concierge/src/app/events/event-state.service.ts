import { Injectable } from '@angular/core';
import { MatDialog } from '@angular/material/dialog';
import { Router } from '@angular/router';
import {
    Booking,
    GroupEventDetailsModalComponent,
    queryBookings,
    removeBooking,
} from '@placeos/bookings';
import {
    AsyncHandler,
    SettingsService,
    notifyError,
    notifySuccess,
    openConfirmModal,
} from '@placeos/common';
import { OrganisationService } from '@placeos/organisation';
import { endOfDay, getUnixTime, startOfDay } from 'date-fns';
import { BehaviorSubject, combineLatest } from 'rxjs';
import { filter, map, shareReplay, switchMap } from 'rxjs/operators';

export interface GroupEventOptions {
    period: 'week' | 'month';
    date?: number;
    end?: number;
    zone_ids?: string[];
}

@Injectable({
    providedIn: 'root',
})
export class EventStateService extends AsyncHandler {
    private _options = new BehaviorSubject<GroupEventOptions>({
        period: 'week',
    });
    private _poll = new BehaviorSubject(0);
    private _changed = new BehaviorSubject(0);

    public readonly event_list = combineLatest([
        this._org.active_building,
        this._options,
        this._changed,
        this._poll,
    ]).pipe(
        filter(([bld]) => !!bld),
        switchMap(([bld, options]) =>
            queryBookings({
                period_start: getUnixTime(startOfDay(options.date)),
                period_end: getUnixTime(
                    endOfDay(options.end || options.date || Date.now())
                ),
                type: 'group-event',
                zones: options.zone_ids?.join(',') || bld.id,
            })
        ),
        map((list) => list.sort((a, b) => a.date - b.date)),
        shareReplay(1)
    );

    public readonly options = this._options.asObservable();

    public changed() {
        this._changed.next(Date.now());
    }

    public get period() {
        return this._options.getValue()?.period;
    }

    constructor(
        private _settings: SettingsService,
        private _org: OrganisationService,
        private _dialog: MatDialog,
        private _router: Router
    ) {
        super();
    }

    public startPolling(delay = 60 * 1000) {
        this.interval(
            'poll',
            () => (document.hasFocus() ? this._poll.next(Date.now()) : ''),
            delay
        );
        return () => this.stopPolling();
    }

    public stopPolling() {
        this.clearInterval('poll');
    }

    public setOptions(options: Partial<GroupEventOptions>) {
        this._options.next({ ...this._options.value, ...options });
    }

    public viewEvent(event: Booking) {
        const ref = this._dialog.open(GroupEventDetailsModalComponent, {
            data: { booking: event, concierge: true },
        });
        this.subscription(
            `edit:${event.id}`,
            ref.componentInstance.edit.subscribe(() => {
                this._router.navigate([
                    '/entertainment',
                    'events',
                    'manage',
                    event.id,
                ]);
            })
        );
        this.subscription(
            `remove:${event.id}`,
            ref.componentInstance.remove.subscribe(async () => {
                await this.removeEvent(event);
                ref.close();
            })
        );
    }

    public async removeEvent(event: Booking) {
        const result = await openConfirmModal(
            {
                title: 'Delete Event',
                content: `Are you sure you want to delete the event "${event.title}"?`,
                icon: { content: 'delete' },
                confirm_text: 'Delete',
            },
            this._dialog
        );
        if (result.reason !== 'done') return;
        result.loading('Deleting event...');
        await removeBooking(event.id)
            .toPromise()
            .catch((e) => {
                notifyError(e);
                result.close();
                throw e;
            });
        result.close();
        notifySuccess('Successfully deleted event.');
        this._changed.next(Date.now());
    }
}
