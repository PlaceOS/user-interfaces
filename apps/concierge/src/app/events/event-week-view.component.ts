import { Component } from '@angular/core';
import { AsyncHandler } from '@placeos/common';
import { EventStateService } from './event-state.service';
import { addDays, format, startOfMinute } from 'date-fns';
import { map, shareReplay, startWith } from 'rxjs/operators';
import { Booking, GroupEventDetailsModalComponent } from '@placeos/bookings';
import { MatDialog } from '@angular/material/dialog';
import { Router } from '@angular/router';

@Component({
    selector: 'event-week-view',
    template: `
        <div class="absolute inset-0 overflow-auto flex">
            <div
                class="sticky left-0 w-24 min-h-full flex flex-col items-end z-20 bg-base-100"
            >
                <div
                    header
                    class="sticky top-0 h-16 min-h-16 flex justify-end z-10"
                >
                    <div class="bg-base-100 pt-6 h-10">
                        <div class="opacity-30 text-xs px-2">
                            {{ now | date: 'zzzz' }}
                        </div>
                    </div>
                </div>
                <div
                    *ngFor="let hour of hours; let i = index"
                    class="relative min-h-10 w-full flex-1 bg-base-100"
                >
                    <div
                        hour
                        class="absolute top-0 right-1 -translate-y-1/2 whitespace-nowrap text-xs opacity-60"
                    >
                        {{ hour }} {{ i >= 12 ? 'PM' : 'AM' }}
                    </div>
                </div>
            </div>
            <div class="relative min-w-[84rem] min-h-full z-10">
                <div
                    header
                    class="sticky flex top-0 h-16 min-h-16 border-b border-base-200 bg-base-100 z-10"
                >
                    <div
                        *ngFor="let date of days"
                        date
                        class="flex flex-col items-center justify-center p-4 min-w-48 flex-1 border-r border-base-200"
                    >
                        <div class="text-sm opacity-60">
                            {{ date | date: 'EEEE' }}
                        </div>
                        <div>{{ date | date: 'd' }}</div>
                    </div>
                </div>
                <div
                    *ngFor="let hour of hours; let i = index"
                    class="relative flex min-h-10 min-w-[84rem] flex-1 border-x border-b border-base-200 pointer-events-none"
                ></div>
                <div
                    *ngFor="let date of days; let i = index"
                    date
                    class="absolute top-16 left-0 w-[calc(100%/7)] h-[60rem] flex-1 border-r border-base-200 pointer-events-none"
                    [style.transform]="'translateX(' + i * 100 + '%)'"
                >
                    <button
                        *ngFor="
                            let event of (event_day_map | async)[
                                dateString(date)
                            ] || []
                        "
                        matRipple
                        class="absolute inset-x-1 bg-base-100 rounded border border-base-200 hover:border-info shadow pl-3 pr-2 py-1 overflow-hidden pointer-events-auto"
                        [style.top]="event.offset * 100 + '%'"
                        [style.height]="event.length * 100 + '%'"
                        (click)="viewEvent(event)"
                    >
                        <div
                            class="absolute left-0 inset-y-0 bg-info w-1.5"
                        ></div>
                        <div class="text-sm opacity-60 h-full text-left">
                            {{ event.date | date: 'shortTime' }} &mdash;
                            {{ event.title }}
                        </div>
                        <div
                            class="absolute inset-0"
                            customTooltip
                            [content]="event_card"
                            xPosition="center"
                            yPosition="bottom"
                            [hover]="true"
                        ></div>
                        <ng-template #event_card>
                            <div class="p-2 pointer-events-none">
                                <group-event-card
                                    [event]="event"
                                ></group-event-card>
                            </div>
                        </ng-template>
                    </button>
                    <div
                        now
                        *ngIf="isCurrentDay(date)"
                        class="absolute inset-x-0 h-0 border-y-2 border-base-content"
                        [style.top]="now_offset * 100 + '%'"
                    >
                        <div
                            class="absolute left-0 top-0 -translate-x-1/2 -translate-y-1/2 h-4 w-4 rounded-full bg-base-content"
                        ></div>
                    </div>
                </div>
            </div>
        </div>
    `,
    styles: [``],
})
export class EventWeekViewComponent extends AsyncHandler {
    public days = new Array(7).fill(0).map((_, idx) => idx + 1);
    public readonly hours = new Array(24)
        .fill(0)
        .map((_, idx) => (idx % 12 ? idx % 12 : 12));
    public readonly event_list = this._state.event_list;
    public readonly event_day_map = this.event_list.pipe(
        map((list) => {
            const map = {};
            for (const event of list) {
                const date = format(event.date, 'yyyy-MM-dd');
                if (!map[date]) map[date] = [];
                const start = new Date(event.date);
                map[date].push({
                    ...event,
                    offset:
                        (start.getHours() * 60 + start.getMinutes()) /
                        (24 * 60),
                    length: event.duration / (24 * 60),
                });
            }
            return map;
        }),
        startWith({}),
        shareReplay(1)
    );

    public readonly viewEvent = (event: any) => this._state.viewEvent(event);

    public dateString(date: number) {
        if (!date) return '';
        return format(date, 'yyyy-MM-dd');
    }

    public isCurrentDay(date: number) {
        if (!date) return false;
        return format(Date.now(), 'yyyy-MM-dd') === format(date, 'yyyy-MM-dd');
    }

    public get now() {
        return startOfMinute(Date.now()).valueOf();
    }

    public get now_offset() {
        const now = new Date(this.now);
        return (now.getHours() * 60 + now.getMinutes()) / (24 * 60);
    }

    constructor(
        private _state: EventStateService,
        private _dialog: MatDialog,
        private _router: Router
    ) {
        super();
    }

    public ngOnInit() {
        this.subscription(
            'date',
            this._state.options.subscribe(({ date }) => {
                if (!date) return;
                this.days = this.days.map((_, idx) =>
                    addDays(date, idx).valueOf()
                );
            })
        );
    }
}
