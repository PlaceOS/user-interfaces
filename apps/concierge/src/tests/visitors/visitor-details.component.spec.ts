import { MatTooltipModule } from '@angular/material/tooltip';
import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';
import { SettingsService } from '@placeos/common';
import { ActionIconComponent, IconComponent } from '@placeos/components';
import { CalendarEvent } from '@placeos/events';
import { GuestUser } from '@placeos/users';
import { MockComponent } from 'ng-mocks';
import { of } from 'rxjs';

import { VisitorDetailsComponent } from '../../app/visitors/visitor-details.component';
import { VisitorsStateService } from '../../app/visitors/visitors-state.service';

jest.mock('@placeos/ts-client');

import * as event_mod from '@placeos/events';
import * as ts_client from '@placeos/ts-client';

describe('VisitorDetailsComponent', () => {
    let spectator: Spectator<VisitorDetailsComponent>;
    const createComponent = createComponentFactory({
        component: VisitorDetailsComponent,
        providers: [
            {
                provide: VisitorsStateService,
                useValue: {
                    checkGuestOut: jest.fn(async () => ({})),
                    checkGuestIn: jest.fn(async () => ({})),
                },
            },
            { provide: SettingsService, useValue: { get: jest.fn() } },
        ],
        declarations: [
            MockComponent(ActionIconComponent),
            MockComponent(IconComponent),
        ],
        imports: [MatTooltipModule],
    });

    beforeEach(() => (spectator = createComponent()));

    it('should create component', () => {
        expect(spectator.component).toBeTruthy();
    });

    it('should match snapshot', () => {
        (ts_client as any).get = jest.fn(() => ({}));
        expect(spectator.element).toMatchSnapshot();
        spectator.setInput({
            event: new CalendarEvent({ date: 1 }),
            visitor: new GuestUser(),
        });
        spectator.detectChanges();
        expect(spectator.element).toMatchSnapshot();
    });

    it('should allow toggling remote state', () => {
        (ts_client as any).get = jest.fn(() => of({}));
        (ts_client as any).patch = jest.fn(() => of({}));
        (ts_client as any).post = jest.fn(() => of({}));
        jest.spyOn(event_mod, 'saveEvent');
        spectator.setInput({
            event: new CalendarEvent(),
            visitor: new GuestUser(),
        });
        spectator.detectChanges();
        spectator.click('action-icon[remote]');
        expect(event_mod.saveEvent).toBeCalled();
    });

    it('should checking in visitor', () => {
        const service = spectator.inject(VisitorsStateService);
        spectator.setInput({
            event: new CalendarEvent(),
            visitor: new GuestUser(),
        });
        spectator.detectChanges();
        spectator.click('action-icon[checkin]');
        expect(service.checkGuestIn).toBeCalled();
    });

    it('should checking out visitor', () => {
        const service = spectator.inject(VisitorsStateService);
        spectator.setInput({
            event: new CalendarEvent(),
            visitor: new GuestUser({ checked_in: true }),
        });
        spectator.detectChanges();
        spectator.click('action-icon[checkout]');
        expect(service.checkGuestOut).toBeCalled();
    });
});
