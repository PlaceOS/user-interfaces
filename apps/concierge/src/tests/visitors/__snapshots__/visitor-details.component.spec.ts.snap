// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VisitorDetailsComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <div
    class="w-12 text-lg flex justify-center"
  >
    <i
      class="p-2 rounded-full material-icons border-2 border-dotted border-gray-600"
    >
      face
    </i>
    
    
    
    
    
  </div>
  <div
    class="p-2 flex-1"
    flex=""
  >
    
  </div>
  <div
    class="w-48 py-2 flex items-center justify-end"
  >
    <action-icon
      aria-describedby="cdk-describedby-message-8"
      cdk-describedby-host=""
      class="mat-tooltip-trigger invisible"
      remote=""
    />
    
    <action-icon
      aria-describedby="cdk-describedby-message-5"
      cdk-describedby-host=""
      checkin=""
      class="mat-tooltip-trigger invisible"
      content="event_available"
      mattooltip="Checkin Guest"
    />
    
    <action-icon
      aria-describedby="cdk-describedby-message-6"
      cdk-describedby-host=""
      checkout=""
      class="mat-tooltip-trigger invisible"
      content="event_busy"
      mattooltip="Checkout Guest"
    />
    
    <a
      aria-describedby="cdk-describedby-message-9"
      cdk-describedby-host=""
      class="mat-tooltip-trigger"
      href="mailto:null"
      mat-icon-button=""
    >
      <app-icon>
        email
      </app-icon>
    </a>
    
    <action-icon
      aria-describedby="cdk-describedby-message-7"
      cdk-describedby-host=""
      class="mat-tooltip-trigger invisible"
      content="event_busy"
      mattooltip="Print QR Code"
    />
    
  </div>
  <div
    class="w-14 p-2"
  />
  
</div>
`;

exports[`VisitorDetailsComponent should match snapshot 2`] = `
<div
  id="root1"
>
  <div
    class="w-12 text-lg flex justify-center"
  >
    <i
      class="p-2 rounded-full material-icons border-2 border-dotted border-gray-600"
    >
      face
    </i>
    
    
    
    
    
  </div>
  <div
    class="p-2 flex-1"
    flex=""
  >
    
  </div>
  <div
    class="w-48 py-2 flex items-center justify-end"
  >
    <action-icon
      aria-describedby="cdk-describedby-message-8"
      cdk-describedby-host=""
      class="mat-tooltip-trigger"
      remote=""
    />
    
    <action-icon
      aria-describedby="cdk-describedby-message-5"
      cdk-describedby-host=""
      checkin=""
      class="mat-tooltip-trigger"
      content="event_available"
      mattooltip="Checkin Guest"
    />
    
    <action-icon
      aria-describedby="cdk-describedby-message-6"
      cdk-describedby-host=""
      checkout=""
      class="mat-tooltip-trigger"
      content="event_busy"
      mattooltip="Checkout Guest"
    />
    
    <a
      aria-describedby="cdk-describedby-message-9"
      cdk-describedby-host=""
      class="mat-tooltip-trigger"
      href="mailto:"
      mat-icon-button=""
    >
      <app-icon>
        email
      </app-icon>
    </a>
    
    <action-icon
      aria-describedby="cdk-describedby-message-7"
      cdk-describedby-host=""
      class="mat-tooltip-trigger invisible"
      content="event_busy"
      mattooltip="Print QR Code"
    />
    
  </div>
  <div
    class="w-14 p-2"
  />
  
</div>
`;
